src/analysis_and_optimization/Mem_pattern.ml

```diff
@@ -0,0 +1,602 @@
+      Set.Poly.singleton decl_id
+  | _ -> Set.Poly.empty
+
+(**
+ * Modify an index's inner expressions with `op`
+ * @param op a functor returning an expression
+ * @param ind the Index.t to modify
+ *)
+let modify_index op ind =
+  match ind with
+  | Index.All -> Index.All
+  | Single ind_expr -> Single (op ind_expr)
+  | Upfrom ind_expr -> Upfrom (op ind_expr)
+  | Between (expr_top, expr_bottom) -> Between (op expr_top, op expr_bottom)
+  | MultiIndex exprs -> MultiIndex (op exprs)
+
+(**
+ * Apply an op returning a Set of strings for `Index` types
+ *  with inner expressions
+ * @param op a functor returning a Set of strings
+ * @param ind the Index.t to modify
+ *)
+let map_index op ind =
+  match ind with
+  | Index.All -> Set.Poly.empty
+  | Single ind_expr -> op ind_expr
+  | Upfrom ind_expr -> op ind_expr
+  | Between (expr_top, expr_bottom) ->
+      Set.Poly.union (op expr_top) (op expr_bottom)
+  | MultiIndex exprs -> op exprs
+
+(**
+ * Apply a functor to an `Index` type's inner returning a boolean if 
+ * the conditional is true for any of the inner expressions.
+ *)
+let name_exists_in_index op ind =
+  match ind with
+  | Index.All -> false
+  | Single ind_expr -> op ind_expr
+  | Upfrom ind_expr -> op ind_expr
+  | Between (expr_top, expr_bottom) -> op expr_top || op expr_bottom
+  | MultiIndex exprs -> op exprs
+
+(**
+ * Search through an expression for `Var name` where `name` is in
+ *  the `aos_exits`
+ * and the type of the expression contains an eigen type.
+ * @param aos_exits A Set of strings with the name of the obj
+ *  we are searching for.
+ * @param pattern An expression pattern we recursivly search through 
+ *  for the name.
+ **)
+let rec query_expr_eigen_names (aos_exits : string Set.Poly.t)
+    Expr.Fixed.({pattern; meta= Expr.Typed.Meta.({type_; _})}) =
+  let query_name = query_expr_eigen_names aos_exits in
+  match pattern with
+  | FunApp ((_ : Fun_kind.t), (exprs : Expr.Typed.Meta.t Expr.Fixed.t list)) ->
+      List.exists ~f:query_name exprs
+  | TernaryIf (predicate, texpr, fexpr) ->
+      query_name predicate || query_name texpr || query_name fexpr
+  | Indexed (expr, indexed) ->
+      query_name expr
+      || List.exists ~f:(name_exists_in_index query_name) indexed
+  | Var (name : string) when UnsizedType.contains_eigen_type type_ ->
+      Set.Poly.mem aos_exits name
+  | Var (_ : string) -> false
+  | Lit ((_ : Expr.Fixed.Pattern.litType), (_ : string)) -> false
+  | EAnd (lhs, rhs) | EOr (lhs, rhs) -> query_name lhs || query_name rhs
+
+(**
+ * Search through an expression for `Var name` where `name` is in
+ *  the `aos_exits` and return `true` if it exists and false otherwise.
+ * and the type of the expression contains an eigen type.
+ * @param aos_exits A Set of strings with the name of the obj
+ *  we are searching for.
+ * @param pattern An expression pattern we recursivly search through 
+ *  for the name.
+ **)
+let rec query_all_expr_eigen_names
```

Comment:
I think this is the same as Mir_utils.expr_var_set

Reply:
Indeed! Though I had to do some other stuff with it to pull out just the eigen types